desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+        # intersection points were evenly spaced, a composite Simpson's quadrature
+        # would require computing the spline on 1.8x more knots for the same accuracy.
+        primitive = polyval(intersect, polyint(integrand)).reshape(
+            pitch.size * M * L, N * R
+        )
+        sums = jnp.cumsum(
+            # Periodic boundary to compute bounce integrals of particles
+            # trapped outside this snapshot of the field lines.
+            jnp.diff(primitive, axis=-1, append=primitive[..., 0, jnp.newaxis])
+            # Multiply by mask that is false at knots of piecewise spline
+            # to avoid adding difference between primitives of splines at knots.
+            * jnp.append(jnp.arange(1, N * R) % R != 0, True),
+            axis=-1,
+        )
+        return fun((sums, is_bp)).reshape(pitch.size, M, L, N * R // 2)
+
+    def _quad_sin(name, pitch=None):
+        """Compute the bounce integral of the named quantity.
+
+        Parameters
+        ----------
+        name : ndarray
+            Name of quantity in ``data_index`` to compute the bounce integral of.
+        pitch : ndarray
+            λ values to evaluate the bounce integral at.
+            If None, uses the values given to the parent function.
+
+        Returns
+        -------
+        F : ndarray, shape(pitch, alpha, rho, (resolution - 1) * 2)
+            The last axis iterates through every bounce integral performed
+            along that field line padded by zeros.
+
+        """
+        pitch, intersect, _, bp1, bp2 = _compute_bp_if_given_pitch(
+            pitch, zeta, poly_B, poly_B_z, *original, err=True
+        )
+        intersect = intersect.reshape(pitch.size * M * L, N * NUM_ROOTS)
+        bp1 = fun((intersect, bp1))
+        bp2 = fun((intersect, bp2))
+        X = x * (bp2 - bp1)[..., jnp.newaxis] + bp2[..., jnp.newaxis]
+        assert X.shape == (pitch.size * M * L, N * 2, x.size)
+
+        def body(i, integral):
+            k = i % (N * 2)
+            j = i // (N * 2)
+            p = i // (M * L * N * 2)
+            v = j % pitch.size
+            # TODO: Add Hermite spline to interpax to pass in B_z[i].
+            integrand = interp1d(X[j, k], zeta, f[v]) / (
+                interp1d(X[j, k], zeta, B_sup_z[v])
+                * jnp.sqrt(1 - pitch[p] * interp1d(X[j, k], zeta, B[v]))
+            )
+            integral = put(integral, i, jnp.sum(w * integrand))
+            return integral
+
+        f = eq.compute(name, grid=grid, override_grid=False, data=data)[name].reshape(
+            M * L, resolution
+        )
+        B_sup_z = data["B^zeta"].reshape(M * L, resolution)
+        return jnp.nan_to_num(
+            # TODO: Vectorize interpax to do this with 1 call with einsum.
+            fori_loop(0, pitch.size * M * L * N * 2, body, jnp.empty(X.shape[:-1]))
+            * jnp.pi
+            / (bp2 - bp1),
+            posinf=0,
+            neginf=0,
+        ).reshape(pitch.size, M, L, N * 2)
+
```

Comment:
minor organizational point, I think it's clearer to define the nested functions _after_ all the constants and variables are defined, so that when reading top to bottom you know what they are when referenced in the nested funcs.